crates/ruff/src/checkers/ast/mod.rs

```diff
@@ -388,12 +390,23 @@ where
                         // be "foo.bar". Given `from foo import bar as baz`, `name` would be "baz"
                         // and `qualified_name` would be "foo.bar".
                         let name = alias.asname.as_ref().unwrap_or(&alias.name);
-                        let qualified_name =
-                            helpers::format_import_from_member(level, module, &alias.name);
+
+                        // Attempt to resolve any relative imports; but if we don't know the current
+                        // module path, or the relative import extends beyond the package root,
+                        // fallback to a literal representation (e.g., `[".", "foo"]`).
+                        let call_path = self
+                            .module_path
+                            .and_then(|module_path| {
+                                resolve_import_from(module_path, level, module, &alias.name)
+                            })
+                            .unwrap_or_else(|| {
+                                collect_import_from_member(level, module, &alias.name)
```

Comment:
I'm not thrilled about this. Basically, if you have `from ...foo import bar`, and the triple dots means you've extended past the module root that we detected for the file, then we can't resolve the import. In the past, we dealt with this in `resolve_call_path` (we'd just return `None` for such imports), but now we need a way to track these in the semantic model, to know that they're unused. So we need _some_ call path for these. The fallback here is that we create a call path like `[".", ".", ".", "foo", "bar"]` in such cases.